### PR TITLE
PARQUET-818: Refactoring to utilize common IO, buffer, memory management abstractions and implementations

### PR DESCRIPTION
**Pull request by wesm**
This refactoring is a bit of a bloodbath, but I've attempted to preserve as much API backwards compatibility as possible. 

Several points

* Arrow does not use exceptions, so will need to be very careful about making sure that no Status goes unchecked. I've tried to get most of them, but might have missed some
* parquet-cpp still exposes an abstract file read and write API as before, but this makes it easy to pass in an Arrow file handle (e.g. HDFS, OS files, memory maps, etc.)
* Custom memory allocators will need to subclass `arrow::MemoryPool` instead. If this becomes onerous for some reason, we can try to find alternatives, but basically it's the exact same class as `parquet::MemoryAllocator`

Does not require any upstream changes in Arrow.

### CODE REVIEW

**Comment by wesm** [comment]
No longer WIP. I got the build and unit tests passing again locally (so painful). Need to do a sweet of the codebase to make sure all buffer resizes and other possible errors are being checked appropriately

**Comment by wesm** [comment]
@xhochy I added Arrow to the thirdparty deps in the README, can you handle adding to the thirdparty build toolchain in #207 after rebasing?

File: src/parquet/arrow/io.h [hunk 1]
```diff
@@ -36,65 +37,25 @@ namespace arrow {
 // existing Arrow memory pool. This way we can direct all allocations to a
 // single place rather than tracking allocations in different locations (for
 // example: without utilizing parquet-cpp's default allocator)
-class PARQUET_EXPORT ParquetAllocator : public MemoryAllocator {
+class PARQUET_EXPORT ParquetAllocator : public MemoryPool {
```
Comment (xhochy) [inline src/parquet/arrow/io.h]: Is there a real need for this? Can't we just pass the plain MemoryPool to Parquet?

File: src/parquet/arrow/utils.h [hunk 2]
```diff
@@ -24,31 +24,7 @@
 #include "parquet/exception.h"
```
Comment (xhochy) [inline src/parquet/arrow/utils.h]: Remove this file, seems no longer needed.

File: src/parquet/util/memory.h [hunk 3]
```diff
@@ -0,0 +1,438 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (xhochy) [inline src/parquet/util/memory.h]: I have the feeling that we could later on move more of this file into Arrow.

**Comment by xhochy** [comment]
@wesm #207 is ready for review/merge. Not sure what needs to be done there besides removing the conditional around Arrow.

**Comment by xhochy** [comment]
Also I really like the ratio of lines changed vs lines removed here. Shows nicely the benefit of using Arrow.

File: src/parquet/arrow/io.h (hunk 1 above)
Comment (wesm) [inline src/parquet/arrow/io.h]: good point, removed

**Comment by wesm** [comment]
OK, this should be ready once the build is green. @majetideepak if there are any problems here that will impact your use of parquet-cpp as a library, can you open follow up JIRAs? Thanks!

**Comment by wesm** [comment]
This is failing because of rpath/LD_LIBRARY_PATH issues. I will statically link Arrow in the unit tests to make this go away

File: src/parquet/file/reader-internal.h [hunk 4]
```diff
@@ -103,8 +103,8 @@ class SerializedFile : public ParquetFileReader::Contents {
   // This class does _not_ take ownership of the data source. You must manage its
   // lifetime separately
   static std::unique_ptr<ParquetFileReader::Contents> Open(
-      std::unique_ptr<RandomAccessSource> source,
-      ReaderProperties props = default_reader_properties());
+      const std::shared_ptr<RandomAccessSource>& source,
```
Comment (majetideepak) [inline src/parquet/file/reader-internal.h]: What is the motivation for making this a `shared_ptr` ?

Comment (wesm) [inline src/parquet/file/reader-internal.h]: I can change it back, but semantically, why would we require that the file reader have exclusive access to a resource? You can always create a wrapper object that wraps a shared_ptr to some other IO interface (e.g., we could have `std::unique_ptr<ArrowInputFile>`, wrapping a shared ptr to one of the Arrow IO implementations).

Comment (wesm) [inline src/parquet/file/reader-internal.h]: Reverted the change, I agree that changing the pointer type in this patch is out of scope

**Comment by wesm** [comment]
Added a static/shared linkage flag for the Arrow libraries. You still need the Arrow headers to build against libparquet, but it at least should simplify linking.

**Comment by wesm** [comment]
@xhochy if you wouldn't mind taking a peek at the change I've made since your review, then this can be merged

File: CMakeLists.txt [hunk 5]
```diff
@@ -692,12 +413,47 @@ if (${CLANG_TIDY_FOUND})
...
+  if(NOT PARQUET_BUILD_STATIC)
+    message(SEND_ERROR "Coverage requires the static lib to be built")
+  endif()
+endif()
+
+#############################################################
+# Apache Arrow linkage
+
+if ("${PARQUET_ARROW_LINKAGE}" STREQUAL "shared")
+  set(ARROW_LINK_LIBS
```
Comment (xhochy) [inline CMakeLists.txt]: This defines the overall linkage for both `parquet_shared` and `parquet_static` but  I would expect that with `PARQUET_ARROW_LINKAGE=shared`, `parquet_static` cannot be linked successfully.

Comment (xhochy) [inline CMakeLists.txt]: Ah, as there is no actual linking happening, this is no problem.

Comment (wesm) [inline CMakeLists.txt]: right -- the static linking is only creating an archive of object code

```
/usr/bin/cmake -E cmake_link_script CMakeFiles/parquet_static.dir/link.txt --verbose=1
ccache /usr/bin/ar cr debug/libparquet.a  CMakeFiles/parquet_objlib.dir/src/parquet/exception.cc.o CMakeFiles/parquet_objlib.dir/src/parquet/types.cc.o CMakeFiles/parquet_objlib.dir/src/parquet/column/levels.cc.o CMakeFiles/parquet_objlib.dir/src/parquet/column/reader.cc.o CMakeFiles/parquet_objlib.dir/src/parquet/column/writer.cc.o CMakeFiles/parquet_objlib.dir/src/parquet/column/scanner.cc.o CMakeFiles/parquet_objlib.dir/src/parquet/column/scan-all.cc.o CMakeFiles/parquet_objlib.dir/src/parquet/column/statistics.cc.o CMakeFiles/parquet_objlib.dir/src/parquet/compression/codec.cc.o CMakeFiles/parquet_objlib.dir/src/parquet/compression/brotli-codec.cc.o CMakeFiles/parquet_objlib.dir/src/parquet/compression/snappy-codec.cc.o CMakeFiles/parquet_objlib.dir/src/parquet/compression/gzip-codec.cc.o CMakeFiles/parquet_objlib.dir/src/parquet/file/metadata.cc.o CMakeFiles/parquet_objlib.dir/src/parquet/file/reader.cc.o CMakeFiles/parquet_objlib.dir/src/parquet/file/reader-internal.cc.o CMakeFiles/parquet_objlib.dir/src/parquet/file/writer.cc.o CMakeFiles/parquet_objlib.dir/src/parquet/file/writer-internal.cc.o CMakeFiles/parquet_objlib.dir/src/parquet/schema/converter.cc.o CMakeFiles/parquet_objlib.dir/src/parquet/schema/descriptor.cc.o CMakeFiles/parquet_objlib.dir/src/parquet/schema/printer.cc.o CMakeFiles/parquet_objlib.dir/src/parquet/schema/types.cc.o CMakeFiles/parquet_objlib.dir/src/parquet/util/cpu-info.cc.o CMakeFiles/parquet_objlib.dir/src/parquet/util/memory.cc.o
ccache /usr/bin/ranlib debug/libparquet.a
```

**Comment by wesm** [comment]
thanks all, merging

**Comment by wesm** [comment]
That felt good.
